hera_cal/utils.py

```diff
@@ -731,3 +731,114 @@ def synthesize_ant_flags(flags, threshold=0.0):
+        by_antpol: if True, chisq and nObs are dictionaries mapping antenna polarizations to numpy arrays.
+            Additionally, if by_antpol is True, cross-polarized visibilities are ignored.
+        chisq: optional chisq to update (see below)
+        nObs: optional nObs to update (see below). Must be specified if chisq is specified and must be
+            left as None if chisq is left as None.
+        chisq_per_ant: optional chisq_per_ant to update (see below)
+        nObs_per_ant: optional nObs_per_ant to update (see below). Must have same keys as chisq_per_ant.
+
+    Returns:
+        chisq: numpy array with the same shape each visibility of chi^2 calculated as above. If chisq
+            is provided, this is the sum of the input chisq and the calculated chisq from all unflagged
+            data-to-model comparisons possible given their overlapping baselines. If by_antpol is True,
+            instead returns a dictionary that maps antenna polarization strings to these numpy arrays.
+        nObs: numpy array with the integer number of unflagged data-to-model comparisons that go into
+            each time and frequency of the chisq calculation. If nObs is specified, this updates that
+            with a count of any new unflagged data-to-model comparisons. If by_antpol is True, instead
+            returns a dictionary that maps antenna polarization strings to these numpy arrays.
+        chisq_per_ant: dictionary mapping ant-pol keys like (1,'x') to chisq per antenna, computed as
+            above but keeping i fixed and varying only j. If chisq_per_ant is specified, this adds in
+            new chisq calculations that include this antenna
+        nObs_per_ant: dictionary mapping ant-pol keys like (1,'x') to the integer number of unflagged
+            data-to-model comparisons that go into each time and frequency of the per-antenna chisq
+            calculation. If nObs_per_ant, this is updated to include all new unflagged observations.
+    """
+    # build containers for chisq and nObs if not supplied
+    if chisq is None and nObs is None:
+        if by_antpol:
+            chisq = {}
+            nObs = {}
+        else:
+            chisq = np.zeros(list(data.values())[0].shape, dtype=float)
+            nObs = np.zeros(list(data.values())[0].shape, dtype=int)
+    elif (chisq is None) ^ (nObs is None):
+        raise ValueError('Both chisq and nObs must be specified or nor neither can be.')
+
+    # build containers for chisq_per_ant and nObs_per_ant if not supplied
+    if chisq_per_ant is None and nObs_per_ant is None:
+        chisq_per_ant = {}
+        nObs_per_ant = {}
+    elif (chisq_per_ant is None) ^ (nObs_per_ant is None):
+        raise ValueError('Both chisq_per_ant and nObs_per_ant must be specified or nor neither can be.')
+
+    for bl in data.keys():
+        if model.has_key(bl) and data_wgts.has_key(bl) and (not by_antpol or bl[2][0] == bl[2][1]):
```

Comment:
can you add a comment what the last conditional is doing? the first two are obvious but I don't quite understand why the two pols entries need to be the same?

Reply:
I've added a comment. It's because if you're doing a separate chi^2 for x and y, it doesn't make sense to include xy and yx visibilities. This is meant to more closely match omnical. For now, it doesn't matter; we don't have xy or yx abscal models.